Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -2038,54 +2081,75 @@ void FunctionDefinitionWriter::visit(const Type* type, AST::CallExpression& call
-            { "reverseBits"_s, "reverse_bits"_s },
-            { "round"_s, "rint"_s },
-            { "sign"_s, "__wgslSign"_s },
-            { "unpack2x16snorm"_s, "unpack_snorm2x16_to_float"_s },
-            { "unpack2x16unorm"_s, "unpack_unorm2x16_to_float"_s },
-            { "unpack4x8snorm"_s, "unpack_snorm4x8_to_float"_s },
-            { "unpack4x8unorm"_s, "unpack_unorm4x8_to_float"_s },
+#define EMIT_HELPER(name) \
+    [](HelperGenerator& helperGenerator) { \
+        if (!std::exchange(helperGenerator.didEmit##name, true)) \
+            helperGenerator.emit##name(); \
+        return "__wgsl"#name##_s;\
+    }
+
+#define NOOP_HELPER(name) \
+    [](HelperGenerator&) { return #name##_s; }
+
+        static constexpr std::pair<ComparableASCIILiteral, ASCIILiteral(*)(HelperGenerator&)> directMappings[] {
+            { "acos"_s, EMIT_HELPER(Acos) },
+            { "acosh"_s, EMIT_HELPER(Acosh) },
+            { "asin"_s, EMIT_HELPER(Asin) },
+            { "atanh"_s, EMIT_HELPER(Atanh) },
+            { "atomicCompareExchangeWeak"_s, NOOP_HELPER(__wgslAtomicCompareExchangeWeak) },
+            { "countLeadingZeros"_s, NOOP_HELPER(clz) },
+            { "countOneBits"_s, NOOP_HELPER(popcount) },
+            { "countTrailingZeros"_s, NOOP_HELPER(ctz) },
+            { "dot"_s, NOOP_HELPER(__wgslDot) },
+            { "dot4I8Packed"_s, NOOP_HELPER(__wgslDot4I8Packed) },
+            { "dot4U8Packed"_s, NOOP_HELPER(__wgslDot4U8Packed) },
+            { "dpdx"_s, NOOP_HELPER(dfdx) },
+            { "dpdxCoarse"_s, NOOP_HELPER(dfdx) },
+            { "dpdxFine"_s, NOOP_HELPER(dfdx) },
+            { "dpdy"_s, NOOP_HELPER(dfdy) },
+            { "dpdyCoarse"_s, NOOP_HELPER(dfdy) },
+            { "dpdyFine"_s, NOOP_HELPER(dfdy) },
+            { "extractBits"_s, NOOP_HELPER(__wgslExtractBits) },
+            { "faceForward"_s, NOOP_HELPER(faceforward) },
+            { "firstLeadingBit"_s, NOOP_HELPER(__wgslFirstLeadingBit) },
+            { "firstTrailingBit"_s, NOOP_HELPER(__wgslFirstTrailingBit) },
+            { "frexp"_s, NOOP_HELPER(__wgslFrexp) },
+            { "fwidthCoarse"_s, NOOP_HELPER(fwidth) },
+            { "fwidthFine"_s, NOOP_HELPER(fwidth) },
+            { "insertBits"_s, NOOP_HELPER(__wgslInsertBits) },
+            { "inverseSqrt"_s, EMIT_HELPER(InverseSqrt) },
+            { "log"_s, EMIT_HELPER(Log) },
+            { "log2"_s, EMIT_HELPER(Log2) },
+            { "modf"_s, NOOP_HELPER(__wgslModf) },
+            { "pack2x16snorm"_s, NOOP_HELPER(pack_float_to_snorm2x16) },
+            { "pack2x16unorm"_s, NOOP_HELPER(pack_float_to_unorm2x16) },
+            { "pack4x8snorm"_s, NOOP_HELPER(pack_float_to_snorm4x8) },
+            { "pack4x8unorm"_s, NOOP_HELPER(pack_float_to_unorm4x8) },
+            { "reverseBits"_s, NOOP_HELPER(reverse_bits) },
+            { "round"_s, NOOP_HELPER(rint) },
+            { "sign"_s, NOOP_HELPER(__wgslSign) },
+            { "sqrt"_s, NOOP_HELPER(Sqrt) },
```

Comment:
I think the above line should be either `NOOP_HELPER(__wgslSqrt)` or `EMIT_HELPER(Sqrt)` otherwise the generated MSL has calls to `Sqrt` which does not exist

Reply:
Noticed running https://webgpu.github.io/webgpu-samples/?sample=cornell

Reply:
Thanks for catching, will re-upload